src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala

```diff
@@ -112,43 +113,103 @@ class DataSource(graphFactory: OrientGraphFactory)(implicit executionContext: Ex
    */
-  var elemLocks = TrieMap[Element, ReentrantReadWriteLock]()
+  var wsLocks = TrieMap[WorkspaceName, StampedLock]()
 
-  private def getElemLock(elem: Element): ReentrantReadWriteLock = {
-    elemLocks.getOrElseUpdate(elem, new ReentrantReadWriteLock())
+  private def getLock(ws: WorkspaceName): StampedLock = {
+    wsLocks.getOrElseUpdate(ws, new StampedLock())
   }
+}
 
-  def withWriteLock[T](elem: Element) (op: => T): T = {
-    val wl = getElemLock(elem).writeLock()
-    try {
-      wl.lock()
-      op
-    } finally {
-      wl.unlock()
+class RawlsTransaction(val graph: OrientGraph, dataSource: DataSource)(implicit executionContext: ExecutionContext) {
+  private val rollbackOnly = new AtomicBoolean(false)
+
+  //Count of locks held by this txn. Handles silently upgrading read to write locks if a write is already held.
+  private val readCounts = TrieMap[WorkspaceName, AtomicInteger]().withDefaultValue(new AtomicInteger(0))
+  private val writeCounts =  TrieMap[WorkspaceName, AtomicInteger]().withDefaultValue(new AtomicInteger(0))
+
+  private def acquireReadLock(ws: WorkspaceName) = {
+    if( writeCounts(ws).get > 0 ) {
+      //Silently "upgrade" read locks to write locks if a write lock exists, since we already
+      //have exclusive access.
+      acquireWriteLock(ws)
+    } else if( readCounts(ws).getAndIncrement == 0 ) {
+      RawlsTransaction.getLock(ws).readLock()
     }
   }
 
-  def withReadLock[T](elem: Element) (op: => T): T = {
-    val rl = getElemLock(elem).readLock()
-    try {
-      rl.lock()
-      op
-    } finally {
-      rl.unlock()
+  private def releaseReadLock(ws: WorkspaceName) = {
+    if( writeCounts(ws).get > 0 ) {
+      //Handle our read lock having been silently upgraded.
+      releaseWriteLock(ws)
+    } else {
+      assert(readCounts(ws).get > 0, s"Attempting to release nonexistent read lock on $ws")
+      if( readCounts(ws).decrementAndGet == 0 ) {
+        val unlocked = RawlsTransaction.getLock(ws).tryUnlockRead()
+        assert(unlocked, s"Someone else apparently released read lock on $ws ahead of readCount == 0???")
+      }
+    }
+  }
+
+  private def acquireWriteLock(ws: WorkspaceName) = {
+    assert( readCounts(ws).get == 0, s"Attempting to acquire write lock on $ws which already has a read lock" )
+    if( writeCounts(ws).getAndIncrement == 0 ) {
+      RawlsTransaction.getLock(ws).writeLock()
+    }
+  }
+
+  private def releaseWriteLock(ws: WorkspaceName) = {
+    assert(writeCounts(ws).get > 0, s"Attempting to release nonexistent write lock on $ws")
+    if( writeCounts(ws).decrementAndGet == 0 ) {
+      val unlocked = RawlsTransaction.getLock(ws).tryUnlockWrite()
+      assert(unlocked, s"Someone else apparently released write lock on $ws ahead of writeCount == 0???")
     }
   }
 
-  def withLock[T](elem: Element, writeLock: Boolean)(op: => T): T = {
+  def withFutureWriteLock[T](ws: WorkspaceName) (op: => Future[T]): Future[T] = {
+    acquireWriteLock(ws)
```

Comment:
If you wanted to do the scala.concurrent.blocking right I think you would wrap acquireXXXLock in a Future:
val lockFuture = Future { blocking { acquireWriteLock(ws) } }
lockFuture.flatMap(op).andThen({case _ => releaseWriteLock(ws)})

Then I don't think you need the try because op is wrapped in another future which catches any exceptions. Might want to verify that ;)